### DISCUSSION TITLE
v2: Drop in replacement of con* types, Constrained* classes and current system of defining custom types using annotated-types

### DISCUSSION THREAD

**Discussion by AlexanderPodorov** [pydantic/pydantic#4992] [Question] [answered]
As far as I noticed, currently in pydantic v1 we have 3 ways of defining constrained types:
1. `con*` functions returning the type.
    * Advantage: simple and concise
    * Big disadvantage: type checker/editor unfriendly
2. `Constrained*` classes.
    * Advantage: simple and concise, type checker/editor friendly
    * Disadvantage: Type checker complains if we pass e.g. plain `str` where `ConstrainedStr` is required. This could be the case if    we create a model manually, without JSON parsing.
3. Sometimes we need use pydantic type for validation (e.g. EmailStr). In this case we have to convert `str` into `EmailStr` in order to be type checker friendly.

We could use e.g. `Annotated[str, Field(..., min_length=1, max_length=10)]` to workaround above limitations, but if we need e.g. to validate email, we have to use EmailStr.

Based on above it looks like pydantic v2 and `annotated-types` integration will be able to solve all that issues.
I just have a few concerns here:
1. Will we have something to support `max_digits` and `decimal_places` of `ConstrainedDecimal`? E.g.:
```python
class MyClass(BaseModel):
    length: Annotated[Decimal, DecimalConstraints(max_digits=8, decimal_places=2)]
```
2. Uniqueness checks in contained types:
```python
class Item(BaseModel):
    id: int
    name: str

class MyClass(BaseModel):
    items: Annotated[Sequence[Item], SequenceConstraints(unique_items=True)]
```
3. Email validation, example usage:
```python
class MyClass(BaseModel):
    email: Annotated[str, EmailValidator()]
```
5. Will we be able to use `Predicate`?, e.g.:
```python
class MyClass(BaseModel):
    age: Annotated[int, Predicate(some_func)]
```
Here we assume that `DecimalConstraints`, `SequenceConstraints` and `EmailValidator` are compatible with `annotated-types`.

I think above things would be ideal, because:
1. Logical: python type and validation is separated.
2. Type checker/editor friendly
3. It might be confusing that we have a bunch (`con*` functions, `Constrained*` types, `Field` function and custom types) of ways of doing similar things
4. Allow to pass plain python types, e.g. `int` and `str`, but get the validation by pydantic:
```python
m = MyClass(email='some@email.com') # no type checker complains

m = MyClass(items=[Item(id=1, name='a), Item(id=2, name='b')]) # we can pass any Sequence, without specifying a custom type or `con*` or `Field` use. 
```

**Comment by samuelcolvin** [comment]
Good question, I'm open to suggestions.

The current options (as of `main` now) are:
* `con*` methods currently remain
* we have `PydanticGeneralMetadata` which allows you to pass any valid argument to a validator, e.g. we use it [here](https://github.com/pydantic/pydantic/blob/afff676c03838cfbc948ad8217f969f686d7796a/pydantic/types.py#LL290C17-L290C40) `max_digits` and `decimal_places` - i guess we need to make `PydanticGeneralMetadata` public
* you can still define custom types which implement `__pydantic_validation_schema__` and `__get_pydantic_validation_schema__`

In addition, I'm happy to add more types like `DecimalConstraints` which are only valid as annotations. This would also cover the case of `Predicate`, although I'm not sure what we should call it or whether we should have separate `BeforePredicate`, `AfterPredicate` and `WrapPredicate` to match the different types of [function validator](https://github.com/pydantic/pydantic-core/blob/7d7ca9f1d11b7145df72c70601271ac3c7113c6e/pydantic_core/core_schema.py#L1272-L1462).

**Reply by AlexanderPodorov** [reply]
> `con*` methods currently remain

Yes, that's good. Are you sure that returning `Annotated[..., ...]` will make them type checker friendly?

> we have `PydanticGeneralMetadata` which allows you to pass any valid argument to a validator

Yes, I think it's a good idea to make `PydanticGeneralMetadata` public. Looks good with `max_digits` and `decimal_places`.

> you can still define custom types which implement `__pydantic_validation_schema__` and `__get_pydantic_validation_schema__`

That's great, in my particular use case, I could define my own custom type e.g. `Email`, and then use it like `email: Annotated[str, Email()]`

I also noticed that `conlist` doesn't have `unique_items` parameter anymore, is it ok?

> In addition, I'm happy to add more types like `DecimalConstraints` which are only valid as annotations.

Maybe it's not necessary for now, since we have `PydanticGeneralMetadata`.

> `Predicate`

Yes, I understand that having `BeforePredicate`, `AfterPredicate` and `WrapPredicate` can make it really complex. I generally think we don't even need Model `@validator` to validate custom types. As you mentioned `__pydantic_validation_schema__` should work well there. It looks like this way we can provide better re-usability.

In my approach I'm aimed to:
1. 100% type checker/editor support
2. Explicitness, so I clearly see what validation was added to a field
3. Manually created models should be type checker friendly as well, e.g. I don't need to convert input data to the target type, so instead of `f = Email('some@email.com')` do this: `f = 'some@email.com'`

Do you plan to make several beta releases and several "release candidate" releases for v2? In order to allow people to test it on their projects and identify use cases better. `SQLAlchemy` did the same thing and it was really helpful. 
Thanks!

**Reply by samuelcolvin** [reply]
> Yes, that's good. Are you sure that returning `Annotated[..., ...]` will make them type checker friendly?

Yes, that's the point of `Annotated`, you can try it now and prove that mypy or pyright are happy.

> I also noticed that conlist doesn't have unique_items parameter anymore, is it ok?

See https://github.com/pydantic/pydantic-core/issues/296

> Do you plan to make several beta releases and several "release candidate" releases for v2?

Yes, absolutely.

**Comment by AlexanderPodorov** [comment, marked as answer]
V2 now supports all above use cases. Discussion is closed.
